File: code-samples/messaging/code-snippets/message-store-export.py
from ringcentral import SDK
from datetime import date
import requests
import time

#
# Create a task to export the account messages within March 2023
#
def create_message_store_report() :
    try:
        endpoint = "/restapi/v1.0/account/~/message-store-report"
        bodyParams = {
    	   'dateFrom': "2023-03-01T00:00:00.000Z",
    	   'dateTo': "2023-03-31T23:59:59.999Z"
        }
        response = platform.post(endpoint, bodyParams)
        json = response.json()
        get_message_store_report_task(json.id)
    except Exception as e:
        print (str(e))

#
# Check the task completion status
#
def get_message_store_report_task(taskId):
    try:
        endpoint = "/restapi/v1.0/account/~/message-store-report/" + taskId
        response = platform.get(endpoint)
        jsonObj = response.json()
        print(f'Task creation status: {jsonObj.status}')
        if jsonObj.status == "Completed":
            return get_message_store_report_archive(taskId)
        elif (jsonObj.status == "AttemptFailed" or jsonObj.status == "Failed" or jsonObj.status == "Cancelled"):
            print ("Export message store failed.")
        else:
            time.sleep(10)
            get_message_store_report_task(taskId)
    except Exception as e:
        print (str(e))

#
# When the task is completed, use the task id to get the uri of the report file
#
def get_message_store_report_archive(taskId):
    print("Getting report uri ...")
    try:
        endpoint = "/restapi/v1.0/account/~/message-store-report/"+ taskId +"/archive"
        response = platform.get(endpoint)
        jsonObj = response.json()
        today = date.today()
        dateLog = today.strftime("%Y_%m_%d_%H_%M")
        for i in range( len(jsonObj.records) ):
            fileName = f'message_store_{dateLog}_{i}.zip'
            get_message_store_report_archive_content(jsonObj.records[i].uri, fileName)
    except Exception as e:
        print (str(e))


def get_message_store_report_archive_content(contentUri, fileName):
    uri = platform.create_url(contentUri, False, None, True);
    response = requests.get(uri, stream=True)
    with open(fileName,"wb") as output:
        output.write(response.content)

    print (f'{fileName} file is saved to the local machine.')


# Authenticate a user using a personal JWT token
def login():
    try:
      platform.login( jwt= "SANDBOX_JWT" )
      create_message_store_report()
    except Exception as e:
      print ("Unable to authenticate to platform. Check credentials." + str(e))

File: code-samples/messaging/code-snippets/message-store-export.rb
require 'ringcentral'
require "down"
require "fileutils"

#
# Create a task to export the account messages within March 2023
#
def create_message_store_report()
  begin
    bodyParams = {
      dateFrom: "2023-03-01T00:00:00.000Z",
      dateTo: "2023-03-31T23:59:59.999Z"
    }
    endpoint = "/restapi/v1.0/account/~/message-store-report"
    response = $platform.post(endpoint, payload: bodyParams)
    puts (response.body['id'])
    get_message_store_report_task(response.body['id'])
  rescue StandardError => e
    puts (e)
  end
end

#
# Check the task completion status
#
def get_message_store_report_task(taskId)
  begin
    endpoint = "/restapi/v1.0/account/~/message-store-report/" + taskId
    resp = $platform.get(endpoint)
    taskStatus = resp.body['status']
    puts ("Task creation status: " + taskStatus)
    if taskStatus == "Completed"
      get_message_store_report_archive(taskId)
    elsif (taskStatus == "AttemptFailed" || taskStatus == "Failed" || taskStatus == "Cancelled")
      puts ("Export message store failed.")
    else
      sleep(10)
      get_message_store_report_task(taskId)
    end
  rescue StandardError => e
    puts (e)
  end
end

#
# When the task is completed, use the task id to get the uri of the report file
#
def get_message_store_report_archive(taskId)
  puts "Getting report uri ..."
  begin
    endpoint = "/restapi/v1.0/account/~/message-store-report/"+ taskId +"/archive"
    resp = $platform.get(endpoint)
    length = resp.body['records'].length
    dateLog = Time.now.strftime("%Y_%m_%d_%H_%M")
    for i in (0...length)
	    fileName = "message_store_" + dateLog + "_" + i.to_s + ".zip"
	    get_message_store_report_archive_content(resp.body['records'][i]['uri'], fileName)
    end
  rescue StandardError => e
    puts (e)
  end
end

def get_message_store_report_archive_content(contentUri, fileName)
  begin
    uri = contentUri + "?access_token=" + $platform.token['access_token']
    tempfile = Down.download(uri)
    FileUtils.mv(tempfile.path, fileName)
    puts (fileName + " file is saved to the local machine.")
  rescue StandardError => e
    puts (e)
  end
end


# Authenticate a user using a personal JWT token
def login()
  begin
    $platform.authorize( jwt: "SANDBOX_JWT" )
    create_message_store_report()
  rescue StandardError => e
    puts ("Unable to authenticate to platform. Check credentials." + e.to_s)
  end
end
